src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -54,6 +54,7 @@ public MethodHandle downcallHandle(FunctionDescriptor function, Option... option
         Objects.requireNonNull(options);
         checkHasNaturalAlignment(function);
         LinkerOptions optionSet = LinkerOptions.of(options);
+        optionSet.validateForDowncall(function);
```

Comment:
Does it make sense to have a validation step that is disjoint from the creation of the LinkerOptions? E.g. should LinkerOptions factory just take the descriptor and perform all the required validation in one go?

Reply:
I don't really like that idea, because it requires the descriptor to be specified twice, which opens up the possibility of them going out of sync.

Reply:
I think we should have 2 validation steps, one when creating the option, where we can check if the option itself is valid, and then another when linking, where we check whether the option makes sense for a particular linkage request.

Reply:
I understand that the steps are logically two, and are implemented separately. But if all the code we have is doing:

```
        LinkerOptions optionSet = LinkerOptions.of(options);
        optionSet.validateForDowncall(function);
```

Then, by exposing them as separate steps, we don't gain much, but it might be the case that, at some point later, another client forgets about calling the validation step.

Reply:
Ah, sorry, I miss read what you were saying. I thought the suggestion was to add the validation + descriptor param when creating Linker.Options.

I agree, doing the validation as part of the factory for LinkerOptions would be better.